giraph-block-app-8/src/main/java/org/apache/giraph/block_app/library/prepare_graph/UndirectedConnectedComponents.java

```diff
@@ -352,10 +352,15 @@ Block calculateConnectedComponentSizes(
     Pair<LongWritable, LongWritable> componentToReducePair = Pair.of(
         new LongWritable(), new LongWritable(1));
     LongWritable reusableLong = new LongWritable();
-    return Pieces.reduceAndBroadcast(
-        "CalcConnectedComponentSizes",
+    // This reduce operation is stateless so we can use a single instance
+    BasicMapReduce<LongWritable, LongWritable, LongWritable> reduceOperation =
         new BasicMapReduce<>(
-            LongTypeOps.INSTANCE, LongTypeOps.INSTANCE, SumReduce.LONG),
+            LongTypeOps.INSTANCE, LongTypeOps.INSTANCE, SumReduce.LONG);
+    return Pieces.reduceAndBroadcastWithArrayOfHandles(
+        "CalcConnectedComponentSizes",
+        3137, /* Just using some large prime number */
```

Comment:
I can't come up with a reason why someone would want to change it. This can start having problems only at trillion components which wouldn't work for many other reasons, for tiny ones this few reducers won't add any overhead, and for larger ones which were currently working this is still improvement since reducers are processed on many machines now.